llmfoundry/callbacks/async_eval_callback.py

```diff
@@ -205,79 +229,186 @@ def __init__(
+            # (which is the most important) is evaled after it is written
+            include_end_of_training=False,
         )
 
+        log.info('Initialized AsyncEval callback. Will generate runs at ' +
+                 f'interval {interval}, checking at {check_interval}')
+
+    def _get_checkpoints_and_launch_runs(self, state: State):
+        """Get the latest checkpoint from the training run.
+
+        Args:
+            state: The current state of the training run
+
+        Returns:
+            Returns checkpoints that have not been evaled
+        """
+        checkpointer = None
+        for callback in state.callbacks:
+            if isinstance(callback, CheckpointSaver):
+                checkpointer = callback
+                break
+
+        if not checkpointer:
+            log.warning('No checkpoint saver callback found. Skipping eval')
+            return
+
+        if not checkpointer.saved_checkpoints:
+            log.debug(
+                'No saved checkpoints found on the checkpointer. Skipping eval')
+            return
+
+        found_checkpoints = set(list_remote_objects(
+            self.checkpoint_save_folder))
+
+        if not found_checkpoints:
+            log.debug('No saved checkpoints found yet on remote. Skipping eval')
+            return
+
+        for checkpoint in checkpointer.saved_checkpoints:
+            # Get the part of the path that contains the interval. This is
+            # different for sharded checkpoints (which are saved in a folder)
+            if state.fsdp_elastic_sharded_enabled:
+                # eg {save_folder}/ep0-ba1/.
+                interval_path = Path(checkpoint).parts[-2]
+            else:
+                # eg {save_folder}/ep0-ba1-rank0.pt
+                interval_path = Path(checkpoint).parts[-1]
+
+            interval = get_interval_from_checkpoint(interval_path,
+                                                    self.interval.unit)
+            if interval.value % self.interval.value != 0:
+                log.debug(
+                    f'Checkpoint {checkpoint} ({interval}) is not at an eval interval ({self.interval}), skipping'
+                )
+                continue
+
+            if interval in self.checkpoints_evaled:
+                continue  # Skip checkpoints that have already been evaled
+
+            # Check if the checkpoint is fully uploaded. If not, skip it until upload is complete
+            if state.fsdp_elastic_sharded_enabled:
+                # checkpoint is something like folder/ep0-ba4/__0_0.distcp
+                checkpoint_folder = '/'.join(checkpoint.split('/')[:-1])
+
+                if f'{checkpoint_folder}/.metadata' not in found_checkpoints:
+                    log.debug(
+                        f'Checkpoint {checkpoint} not fully uploaded (missing metadata), skipping'
+                    )
+                    continue
+
+                for i in range(dist.get_world_size()):
```

Comment:
perhaps it would be a bit less fragile to just count the number of files in the folder with the right file ending? im not sure if torch might change these filenames at some point.